converters/fromZigbee.js

```diff
@@ -5450,12 +5450,17 @@ const converters = {
     RTCGQ12LM_occupancy_illuminance: {
         cluster: 'aqaraOpple',
         type: ['attributeReport', 'readResponse'],
-        options: [exposes.options.precision('illuminance'), exposes.options.calibration('illuminance', 'percentual')],
+        options: [exposes.options.precision('illuminance'), exposes.options.calibration('illuminance', 'percentual'),
+            exposes.numeric('motion_wait_timeout').withValueMin(0).withUnit('ms')
+                .withDescription('If a new motion event occurs within the time specified in this option (value is in ms) ' +
+                'after the end of "detection_interval", the "occupancy" state is not cleared')],
         convert: (model, msg, publish, options, meta) => {
             if (msg.data.hasOwnProperty('illuminance')) {
                 // The occupancy sensor only sends a message when motion detected.
                 // Therefore we need to publish the no_motion detected by ourselves.
-                const timeout = meta && meta.state && meta.state.hasOwnProperty('detection_interval') ? meta.state.detection_interval : 60;
+                let timeout = meta && meta.state && meta.state.hasOwnProperty('detection_interval') ? meta.state.detection_interval : 60;
+                const delay = options && options.hasOwnProperty('motion_wait_timeout') ? options.motion_wait_timeout : 0;
+                if (delay > 0) timeout += delay/1000;
```

Comment:
I don't understand this option, why not increase the `detection_interval ` instead?

Reply:
If this option is not used, then regardless of the value of `detection_interval`, and when motion is detected by the sensor, it will first go to the false state, and then immediately to true.

This option is needed to avoid the sensor going to false if it still detects motion after `detection_interval` ends.

Reply:
> #3600
> Compared to this, is there any improvement?

Yes, `motion_wait_timeout` will allow you to do the same as `occupancy_timeout`, but more precisely because the delay is expressed in ms, which may be relevant when setting `detection_interval` = 2s and the need for a minimum delay.

Reply:
As far as I understand, this problem occurs due to the delay in the transmission over the radio channel and the processing of the motion detection message from the sensor, i.e. the _timeout_ equal to `detection_interval` has already ended, and a new message from the sensor has not yet arrived.
_In my case, this delay is about a few tenths of a second._

Here is the log of the **RTCGQ13LM** sensor, which constantly detects movement:

- with `detection_interval` **= 2s** and without using `motion_wait_timeout`
  
  ```javascript
  debug 2022-03-17 01:52:54: Received Zigbee message from '0x04cf8cdf3c8a82e0', type 'attributeReport', cluster 'msOccupancySensing', data '{"occupancy":1}' from endpoint 1 with groupID 0
  info  2022-03-17 01:52:54: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":152,"motion_sensitivity":"high","occupancy":true,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  info  2022-03-17 01:52:56: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":152,"motion_sensitivity":"high","occupancy":false,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  debug 2022-03-17 01:52:56: Received Zigbee message from '0x04cf8cdf3c8a82e0', type 'attributeReport', cluster 'msOccupancySensing', data '{"occupancy":1}' from endpoint 1 with groupID 0
  info  2022-03-17 01:52:56: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":155,"motion_sensitivity":"high","occupancy":true,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  info  2022-03-17 01:52:58: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":155,"motion_sensitivity":"high","occupancy":false,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  debug 2022-03-17 01:52:58: Received Zigbee message from '0x04cf8cdf3c8a82e0', type 'attributeReport', cluster 'msOccupancySensing', data '{"occupancy":1}' from endpoint 1 with groupID 0
  info  2022-03-17 01:52:58: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":152,"motion_sensitivity":"high","occupancy":true,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  info  2022-03-17 01:53:00: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":152,"motion_sensitivity":"high","occupancy":false,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  debug 2022-03-17 01:53:00: Received Zigbee message from '0x04cf8cdf3c8a82e0', type 'attributeReport', cluster 'msOccupancySensing', data '{"occupancy":1}' from endpoint 1 with groupID 0
  info  2022-03-17 01:53:00: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":157,"motion_sensitivity":"high","occupancy":true,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  info  2022-03-17 01:53:02: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":157,"motion_sensitivity":"high","occupancy":false,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  debug 2022-03-17 01:53:02: Received Zigbee message from '0x04cf8cdf3c8a82e0', type 'attributeReport', cluster 'msOccupancySensing', data '{"occupancy":1}' from endpoint 1 with groupID 0
  info  2022-03-17 01:53:02: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":157,"motion_sensitivity":"high","occupancy":true,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  ```

- with `detection_interval` **= 2s** and `motion_wait_timeout` **= 600ms**

  ```javascript
  debug 2022-03-17 02:06:58: Received Zigbee message from '0x04cf8cdf3c8a82e0', type 'attributeReport', cluster 'msOccupancySensing', data '{"occupancy":1}' from endpoint 1 with groupID 0
  info  2022-03-17 02:06:58: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":160,"motion_sensitivity":"high","occupancy":true,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  debug 2022-03-17 02:07:00: Received Zigbee message from '0x04cf8cdf3c8a82e0', type 'attributeReport', cluster 'msOccupancySensing', data '{"occupancy":1}' from endpoint 1 with groupID 0
  info  2022-03-17 02:07:00: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":160,"motion_sensitivity":"high","occupancy":true,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  debug 2022-03-17 02:07:02: Received Zigbee message from '0x04cf8cdf3c8a82e0', type 'attributeReport', cluster 'msOccupancySensing', data '{"occupancy":1}' from endpoint 1 with groupID 0
  info  2022-03-17 02:07:02: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":155,"motion_sensitivity":"high","occupancy":true,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  debug 2022-03-17 02:07:04: Received Zigbee message from '0x04cf8cdf3c8a82e0', type 'attributeReport', cluster 'msOccupancySensing', data '{"occupancy":1}' from endpoint 1 with groupID 0
  info  2022-03-17 02:07:04: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":157,"motion_sensitivity":"high","occupancy":true,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  debug 2022-03-17 02:07:06: Received Zigbee message from '0x04cf8cdf3c8a82e0', type 'attributeReport', cluster 'msOccupancySensing', data '{"occupancy":1}' from endpoint 1 with groupID 0
  info  2022-03-17 02:07:06: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":178,"motion_sensitivity":"high","occupancy":true,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  info  2022-03-17 02:07:08: MQTT publish: topic 'zigbee2mqtt/0x04cf8cdf3c8a82e0', payload '{"battery":100,"detection_interval":2,"linkquality":178,"motion_sensitivity":"high","occupancy":false,"state":"OFF","temperature":25,"update":{"state":null},"update_available":null,"voltage":3123}'
  ```

Reply:
Here's what I understand: 
`occupancy: true` needs `occupancy_timeout` to keep, and `occupancy_timeout` needs `detection_interval` to reset.

Reply:
The latency problem can be solved using `occupancy_timeout` or `motion_wait_timeout`, with the only difference being that `motion_wait_timeout` allows you to set the _timeout to less than a second_.

`occupancy_timeout` is nothing more than `detection_interval` + `motion_wait_timeout`. Everything is simple :-)

Reply:
I think it would be better to add a default delay such that it works out-of-the-box. So e.g. `timeout = detection_interval + 2 seconds`

Reply:
Ok, [changed](https://github.com/Koenkk/zigbee-herdsman-converters/pull/4006/commits/235038d69a0e1ecdf0d08c5e0478da90947de53a) default value of `motion_wait_timeout` to 2s.

Reply:
Thanks, does it still make sense that have an option to override this? (what is the added value of that?)

Reply:
Of course, it should be possible to set any value, because, for example,

`detection_interval` = 60s + `motion_wait_timeout` = 2s
and  
`detection_interval` = 2s + `motion_wait_timeout` = 60s

these are two big differences.

Reply:
I see your point now, could we rename it to `occupancy_timeout` to bring it more in line with others? https://www.zigbee2mqtt.io/devices/RTCGQ11LM.html#options . Note that here you can still express a sub second precision using e.g. `0.1` (=100ms)

Reply:
Initially, I tried to solve the current problem using `occupancy_timeout`, but due to the inability to set a fractional value in the web interface z2m (_input field does not allow you to enter "." character from keyboard_) and the default value is too large (90s), I decided to create my own variant in in the form of `motion_wait_timeout` with the possibility of specifying a value in ms and with a default value of 0.

If we solve the problem with the input field in the web interface and reduce the default values ​​that are too large for **RTCGQ12LM** and **RTCGQ13LM**, then it will be possible to abandon `motion_wait_timeout` in favor of `occupancy_timeout`.

Reply:
That should be a matter of defining a new option: https://github.com/Koenkk/zigbee-herdsman-converters/blob/dc1f12ed3b03b7e490ede8c537e1ffe00c9cc069/lib/exposes.js#L489

e.g.
```js
        occupancy_timeout_2: () => new Numeric(`occupancy_timeout`, access.SET).withValueMin(0).withValueStep(0.1).withDescription('Time in seconds after which occupancy is cleared after detecting it (default 2 seconds).'),

```

Reply:
You can use it directly `[fz.occupancy_with_timeout`

Reply:
> You can use it directly `fz.occupancy_with_timeout`

Unfortunately, `fz.occupancy_with_timeout` in its current state is not really suitable for **RTCGQ13LM** and is not suitable for **RTCGQ12LM** at all.

Reply:
> That should be a matter of defining a new option

[Replaced](https://github.com/Koenkk/zigbee-herdsman-converters/pull/4006/commits/6be8f6b2d39df72012e44022f6473e4cdde24a23) `motion_wait_timeout` with modified `occupancy_timeout` and brought the general logic of work in line with `fz.occupancy_with_timeout`, added `no_occupancy_since_true` and `temperature`.

Reply:
> due to the inability to set a fractional value in the web interface z2m (_input field does not allow you to enter "." character from keyboard_)

During testing, it turned out that the separator of a fractional number is "," :-)

Reply:
great, thanks again!